### PR TITLE
get pd leader from etcd. 

### PR DESCRIPTION
**Pull request by siddontang**
Now etcd supports HTTP JSON API, so we can use hyper to get the pd leader from Etcd now.

The pd flag is deprecated, and we should use etcd flag to start tikv-server like:
```
tikv-server --etcd=127.0.0.1:2379
```
So we will change many places like docker, ci, etc.

@ngaut @BusyJay @disksing 

/cc @iamxy @shenli

### CODE REVIEW

File: travis-build/Makefile [hunk 1]
```diff
@@ -34,12 +34,29 @@ ${KCOV_DIR}/kcov:
 	mkdir -p ${KCOV_DIR} && \
 	cp src/kcov ${KCOV_DIR}/kcov
 
-prepare_linux: ${ROCKSDB_DIR}/librocksdb.so ${KCOV_DIR}/kcov
+# TODO: After Etcd releases binary package, we don't need to install go and then build the source.
```
Comment (siddontang) [inline travis-build/Makefile]: Hi @xiang90 

When do you release next 3.0 beta which supports restful API?

Comment (xiang90) [inline travis-build/Makefile]: 6.30

**Comment by ngaut** [comment]
LGTM

File: src/bin/tikv-server.rs [hunk 2]
```diff
@@ -391,13 +391,20 @@ fn run_local_server(listener: TcpListener, store: Storage, config: &Config) {
...
-                                   None,
-                                   |v| v.as_str().map(|s| s.to_owned()));
-    let pd_client = Arc::new(new_rpc_client(&pd_addr, cfg.cluster_id).unwrap());
+    let etcd_endpoints = get_string_value("etcd",
+                                          "etcd.endpoints",
+                                          matches,
+                                          config,
+                                          None,
+                                          |v| v.as_str().map(|s| s.to_owned()));
+    let etcd_pd_root = get_string_value("etcd-pd",
```
Comment (BusyJay) [inline src/bin/tikv-server.rs]: pd-root seems better.

File: src/pd/etcd.rs [hunk 3]
```diff
@@ -0,0 +1,355 @@
...
+    Next::write().timeout(Duration::from_secs(HTTP_REQUEST_TIMEOUT))
+}
+
+// Encode the key with base64 format and then covert
+// to Json object.
+pub fn b64encode_to_json(key: &[u8]) -> Json {
+    key.to_base64(STANDARD).to_json()
+}
+
+fn parse_response<T: Read>(r: &mut T) -> Result<Json> {
```
Comment (BusyJay) [inline src/pd/etcd.rs]: What's difference between this function and `Json::from_reader`?

File: src/pd/etcd.rs [hunk 4]
```diff
@@ -0,0 +1,355 @@
...
+    let obj = box_try!(Json::from_reader(r));
+    Ok(obj)
+}
+
+pub fn parse_leader_addr(obj: &Json, leader_key: &[u8]) -> Result<String> {
+    if let Some(&Json::Array(ref kvs)) = obj.find("kvs") {
+        if kvs.len() != 1 {
```
Comment (BusyJay) [inline src/pd/etcd.rs]: Will there be more keys in the future?

File: src/pd/etcd.rs [hunk 5]
```diff
@@ -0,0 +1,355 @@
...
+        if let Err(Error::Hyper(_)) = res {
+            // we meet HyperError, maybe we can't communicate with
+            // the etcd server, so try to next one next time.
+            self.next_index = (self.next_index + 1) % self.endpoints.len();
+        }
+
+        res
+    }
+
+    pub fn get_leader_addr(&mut self) -> Result<String> {
```
Comment (BusyJay) [inline src/pd/etcd.rs]: Is it Etcd's leader?

File: travis-build/test.sh [hunk 6]
```diff
@@ -14,6 +14,11 @@
 
 set -o pipefail
 
+# start etcd
+etcd &
```
Comment (BusyJay) [inline travis-build/test.sh]: Do we need to kill it before exit?

Comment (siddontang) [inline travis-build/test.sh]: no need, global is ok.

File: src/pd/etcd.rs (hunk 5 above)
Comment (siddontang) [inline src/pd/etcd.rs]: no need to care,a any is ok.

File: src/pd/etcd.rs (hunk 4 above)
Comment (siddontang) [inline src/pd/etcd.rs]: No, only one.

File: src/pd/etcd.rs (hunk 3 above)
Comment (siddontang) [inline src/pd/etcd.rs]: outer no need to handle json error.

File: src/pd/etcd.rs (hunk 5 above)
Comment (BusyJay) [inline src/pd/etcd.rs]: What I mean is whose leader does this API return?

Comment (siddontang) [inline src/pd/etcd.rs]: pd

File: src/pd/etcd.rs [hunk 7]
```diff
@@ -0,0 +1,355 @@
...
+            cause(err.as_ref())
+            description(err.description())
+            display("unknown error {:?}", err)
+        }
+    }
+}
+
+pub type Result<T> = result::Result<T, Error>;
+
+fn next_read() -> Next {
```
Comment (BusyJay) [inline src/pd/etcd.rs]: `read_next` seems better.

File: src/pd/etcd.rs (hunk 3 above)
Comment (BusyJay) [inline src/pd/etcd.rs]: I think use `try!(parse_response)` and `box_try!(Json::from_reader)` are the same. So no need to wrap another function.

File: src/pd/etcd.rs (hunk 7 above)
Comment (siddontang) [inline src/pd/etcd.rs]: It is for `Next::read()`, I think it is ok.

File: src/pd/etcd.rs (hunk 3 above)
Comment (siddontang) [inline src/pd/etcd.rs]: we can't return directly outer.

File: src/pd/etcd.rs (hunk 5 above)
Comment (BusyJay) [inline src/pd/etcd.rs]: `get_pd_leader_addr` seems better.

File: travis-build/test.sh (hunk 6 above)
Comment (BusyJay) [inline travis-build/test.sh]: I think kill it seems better. So I don't need to kill it myself when I run the script on my machine.

File: src/pd/etcd.rs (hunk 5 above)
Comment (siddontang) [inline src/pd/etcd.rs]: here is just for pd, so it is ok.

Comment (siddontang) [inline src/pd/etcd.rs]: I will rename EtcdClient -> EtcdPdClient for more clear.

File: travis-build/test.sh [hunk 8]
```diff
@@ -33,4 +41,5 @@ for l in sys.stdin:
     grep $case $LOG_FILE | cut -d ' ' -f 2-
     echo
 done
+pkill -9 etcd
```
Comment (BusyJay) [inline travis-build/test.sh]: Use `trap 'kill $(jobs -p)' EXIT` at the beginning instead.

File: src/pd/etcd.rs [hunk 9]
```diff
@@ -0,0 +1,355 @@
...
+        if let Err(e) = res {
+            return Err(box_err!("request {} at {} failed {}", path, endpoint, e));
+        }
+
+        let timeout = Duration::from_secs(HTTP_REQUEST_TIMEOUT);
+        if !finish.wait_timeout(Some(timeout)) {
+            return Err(Error::Timeout(timeout));
+        }
+
+        let res = finish.take().unwrap();
```
Comment (BusyJay) [inline src/pd/etcd.rs]: `finish.take` may return `None`.

**Comment by BusyJay** [comment]
`create_mem_buf` in **conn.rs** can be deleted.

**Comment by siddontang** [comment]
yes, I will update conn after this PR merged.

**Comment by BusyJay** [comment]
Why not do it in this PR?

**Comment by siddontang** [comment]
this PR is for pd, I prefer one thing one PR.

**Comment by BusyJay** [comment]
LGTM

**Pull request merged**